tfe/provider.go

```diff
@@ -0,0 +1,200 @@
+
+	// Add any static host configurations service discovery object.
+	for userHost, hostConfig := range config.Hosts {
+		host, err := svchost.ForComparison(userHost)
+		if err != nil {
+			// ignore invalid hostnames.
+			continue
+		}
+		services.ForceHostServices(host, hostConfig.Services)
+	}
+
+	// Discover the full Terraform Enterprise service address.
+	address := services.DiscoverServiceURL(host, serviceID)
+	if address == nil {
+		return nil, fmt.Errorf("host %s does not provide a Terraform Enterprise API", host)
+	}
+
+	// Only try to get to the token from the credentials source if no token
+	// was explicitly set in the provider configuration.
+	if token == "" {
+		creds, err := services.CredentialsForHost(host)
+		if err != nil {
+			log.Printf("[DEBUG] Failed to get credentials for %s: %s (ignoring)", host, err)
+		}
+		if creds != nil {
+			token = creds.Token()
+		}
+	}
+
+	// If we still don't have a token at this point, we return an error.
+	if token == "" {
+		return nil, fmt.Errorf("required token could not be found")
+	}
+
+	// Create a new TFE client config..
+	cfg := &tfe.Config{
+		Address: address.String(),
+		Token:   token,
+	}
+
+	// Create s new TFE client.
+	return tfe.NewClient(cfg)
+}
+
+// cliConfig tries to find and parse the configuration of the Terraform CLI.
+// This is an optional step, so any errors are ignored.
+func cliConfig() *Config {
+	config := &Config{}
+
+	// Detect the CLI config file path.
+	configFilePath := os.Getenv("TERRAFORM_CONFIG")
+	if configFilePath == "" {
+		filePath, err := configFile()
+		if err != nil {
+			log.Printf("[ERROR] Error detecting default CLI config file path: %s", err)
+			return config
+		}
+		configFilePath = filePath
+	}
+
+	// Read the CLI config file content.
+	content, err := ioutil.ReadFile(configFilePath)
+	if err != nil {
+		log.Printf("[ERROR] Error reading the CLI config file: %v", err)
+		return config
+	}
+
+	// Parse the CLI config file content.
+	obj, err := hcl.Parse(string(content))
+	if err != nil {
+		log.Printf("[ERROR] Error parsing the CLI config file: %v", err)
+		return config
+	}
+
+	// Decode the CLI config file content.
+	if err := hcl.DecodeObject(&config, obj); err != nil {
+		log.Printf("[ERROR] Error decoding the CLI config file: %v", err)
+	}
+
+	return config
+}
+
+func credentialsSource(config *Config) auth.CredentialsSource {
+	creds := auth.NoCredentials
+
+	// Add all configured credentials to the credentials source.
+	if len(config.Credentials) > 0 {
+		staticTable := map[svchost.Hostname]map[string]interface{}{}
+		for userHost, creds := range config.Credentials {
+			host, err := svchost.ForComparison(userHost)
+			if err != nil {
+				// We expect the config was already validated by the time we get
+				// here, so we'll just ignore invalid hostnames.
+				continue
+			}
+			staticTable[host] = creds
+		}
+		creds = auth.StaticCredentialsSource(staticTable)
+	}
+
+	return creds
+}
+
+var descriptions = map[string]string{
```

Comment:
You can just inline these in the resource schema. It ends up being more readable.

Reply:
I know we can, but I see this approach being used a lot as well and personally I like this better (code looks cleaner this way). But if you think inline descriptions is the standard or is preferred, I'll update it. Just let me know...

Reply:
I _do_ think it's standard, and I've never seen anything otherwise, but Terraform is a big codebase, so I could be mistaken there. I also don't think it's a big deal. I think, as more resources are added, we'll find that the indirection doesn't add any value and makes it harder to parse, but because it's not a user-facing change, I think we can resolve that when it happens.

Reply:
Ok... In that case I will update this in the next release.